File: BuyuSources/ServerCore/SourceCode/GameServer/RoleGameData.cpp
#include "Stdafx.h"
#include "RoleGameData.h"
#include "RoleEx.h"
#include "FishServer.h"
RoleGameData::RoleGameData()
{
	m_IsLoadDB = false;
}
RoleGameData::~RoleGameData()
{
	//SaveRoleGameData();
}
bool RoleGameData::OnInit(CRoleEx* pRole)
{
	if (!pRole)
	{
		ASSERT(false);
		return false;
	}
	m_pRole = pRole;
	//发送数据库
	m_IsLoadDB = false;
	OnLoadRoleGameDataByDB();
	return true;
}
void RoleGameData::OnLoadRoleGameDataByDB()
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	/*DBR_Cmd_LoadGameData msg;
	SetMsgInfo(msg, DBR_LoadGameData, sizeof(DBR_Cmd_LoadGameData));
	msg.dwUserID = m_pRole->GetUserID();
	g_FishServer.SendNetCmdToDB(&msg);*/
}
void RoleGameData::OnLoadRoleGameDataResult(DBO_Cmd_LoadGameData* pMsg)
{
	if (!m_pRole || !pMsg)
	{
		ASSERT(false);
		return;
	}
	//将数据库和结构绑定起来
	m_RoleGameData = pMsg->GameData;
	m_IsLoadDB = true;
	m_pRole->IsLoadFinish();
}
void RoleGameData::SendRoleGameDataToClient()
{
	//将数据发送到客户端去
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	LC_Cmd_GendGameData msg;
	SetMsgInfo(msg, GetMsgType(Main_GameData, LC_GendGameData), sizeof(LC_Cmd_GendGameData));
	msg.GameData = m_RoleGameData;
	m_pRole->SendDataToClient(&msg);
}
//void RoleGameData::SaveRoleGameData()
//{
//	//发送数据库命令 将数据保存起来
//	if (!m_IsNeedSave || !m_pRole)
//		return;
//	//发送DBR命令 保存数据
//	DBR_Cmd_SaveGameData msg;
//	SetMsgInfo(msg, DBR_SaveGameData, sizeof(DBR_Cmd_SaveGameData));
//	msg.dwUserID = m_pRole->GetUserID();
//	msg.GameData = m_RoleGameData;
//	g_FishServer.SendNetCmdToDB(&msg);
//	m_IsNeedSave = false;
//}
void RoleGameData::OnHandleCatchFish(BYTE FishID)
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	//玩家捕获指定ID的鱼.
	m_RoleGameData.RoleCatchFishTotalSum += 1;	
	if (FishID == 9)
	{
		m_RoleGameData.CatchFishSum_9 += 1;
	}
	else if (FishID == 18)
	{
		m_RoleGameData.CatchFishSum_18 += 1;
	}
	else if (FishID == 20)
	{
		m_RoleGameData.CatchFishSum_20 += 1;
	}
	else if (FishID == 1)
	{
		m_RoleGameData.CatchFishSum_1 += 1;
	}
	else if (FishID == 3)
	{
		m_RoleGameData.CatchFishSum_3 += 1;
	}
	else if (FishID == 19)
	{
		m_RoleGameData.CatchFishSum_19 += 1;
	}
	m_pRole->SetRoleIsNeedSave();
}
void RoleGameData::OnHandleRoleGetGlobel(DWORD AddGlobel)
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	m_RoleGameData.RoleGetGlobelSum += AddGlobel;
	m_pRole->SetRoleIsNeedSave();
}
void RoleGameData::OnHandleRoleMonthReward(int RewardIndex)
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	m_RoleGameData.RoleMonthRewardSum += 1;
	if (RewardIndex == 1)
		m_RoleGameData.RoleMonthFirstSum += 1;
	else if (RewardIndex == 2)
		m_RoleGameData.RoleMonthSecondSum += 1;
	else if (RewardIndex == 3)
		m_RoleGameData.RoleMonthThreeSum += 1;
	m_pRole->SetRoleIsNeedSave();
}
void RoleGameData::OnHandleRoleJoinTable(bool IsMonth)
{
	m_IsInMonth = IsMonth;
	m_LogJoinTableTime = timeGetTime();
}
void RoleGameData::OnHandleRoleLeaveTable()
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	//当用户离开桌子的时候 
	//1.判断玩家是否离开比赛的桌子 并且 判断玩家在桌子上的游戏时间 用于统计数据
	DWORD GameSec = (timeGetTime() - m_LogJoinTableTime) / 1000;
	m_RoleGameData.TotalGameSec += GameSec;
	if (!m_IsInMonth)
		m_RoleGameData.NonMonthGameSec += GameSec;
	m_pRole->SetRoleIsNeedSave();
}
void RoleGameData::OnHandleRoleSignUpMonth()
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	//玩家报名参加比赛
	m_RoleGameData.RoleMonthSigupSum += 1;
	m_pRole->SetRoleIsNeedSave();
}
void RoleGameData::OnHandleRoleComb()
{
	if (!m_pRole)
	{
		ASSERT(false);
		return;
	}
	CRole* pRole = g_FishServer.GetTableManager()->SearchUser(m_pRole->GetUserID());
	if (!pRole)
	{
		return;
	}
	WORD ComnoSum = pRole->CombCount();
	if (m_RoleGameData.MaxCombSum >= ComnoSum)
		return;
	m_RoleGameData.MaxCombSum = ComnoSum;//记录玩家最大的连击数
	m_pRole->SetRoleIsNeedSave();
}

File: BuyuSources/NewClient/Assets/Scripts/Player/Role/RoleGameData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

class RoleGameData
{
    private bool m_IsLoadInfo = false;
    private tagRoleGameData m_GameData = new tagRoleGameData();
    private bool m_IsLoadByServer = false;
    private bool m_IsInMonth;
    private UInt32 m_LogJoinTableTime;

    public RoleGameData()
    {
        OnClear();
    }
    public void OnClear()
    {
        //清理数据
        m_IsLoadInfo = false;
        m_IsLoadByServer = false;
        m_GameData = new tagRoleGameData();
    }
    public void ResetInfo()
    {
        OnClear();
        tagGameDataChangeEvent pEvent = new tagGameDataChangeEvent();
        MsgEventHandle.HandleMsg(pEvent);
    }
    public tagRoleGameData GetRoleGameData()
    {
        if (m_IsLoadInfo)
        {
            return m_GameData;
        }
        else
        {
            if (!m_IsLoadByServer)
            {
                m_IsLoadByServer = true;
                SendLoadRoleGameData();
            }
            return null;
        }
    }

    //根据服务器端 一些特殊处理的函数
    public void OnHandleCatchFish(Byte FishID)
    {
	    //玩家捕获指定ID的鱼.
        if (!m_IsLoadInfo)
            return;
	    m_GameData.RoleCatchFishTotalSum += 1;
        if (FishID == 9)
        {
            m_GameData.CatchFishSum_9 += 1;
        }
        else if (FishID == 18)
        {
            m_GameData.CatchFishSum_18 += 1;
        }
        else if (FishID == 20)
        {
            m_GameData.CatchFishSum_20 += 1;
        }
        else if (FishID == 1)
        {
            m_GameData.CatchFishSum_1 += 1;
        }
        else if (FishID == 3)
        {
            m_GameData.CatchFishSum_3 += 1;
        }
        else if (FishID == 19)
        {
            m_GameData.CatchFishSum_19 += 1;
        }
    }
    public void OnHandleRoleGetGlobel(UInt32 AddGlobel)
    {
        if (!m_IsLoadInfo)
            return;
	    m_GameData.RoleGetGlobelSum += AddGlobel;
    }

    public string GetTimeStrBySec(UInt32 Sec)
    {
        //根据秒数显示时间  %D %h %m %
        UInt32 Hour = Sec /  3600;
        UInt32 Min = (Sec % 3600) / 60;
        //将4个参数拼接为字符串
        string TimeStr = "";
        if (Hour > 0)
            TimeStr = string.Format("" + Hour.ToString() + "小时+");
        else
            TimeStr = string.Format("" + Min.ToString() + "分钟");
        return TimeStr;
    }
    public string GetGoldStrBygold(UInt32 gold)
    {
        string strGold;
        //100W以上缩为简写
        if (gold >= 1000000)
        {
            gold /= 10000;
            strGold = string.Format("" + gold.ToString() + "万+");
        }
        else
        {
            strGold = gold.ToString();
        }

        return strGold;
    }
    public void OnHandleRoleMonthReward(uint RewardIndex)
    {
        if (!m_IsLoadInfo)
            return;
        m_GameData.RoleMonthRewardSum += 1;
	    if (RewardIndex == 1)
            m_GameData.RoleMonthFirstSum += 1;
	    else if (RewardIndex == 2)
            m_GameData.RoleMonthSecondSum += 1;
	    else if (RewardIndex == 3)
            m_GameData.RoleMonthThreeSum += 1;
    }
    public void OnHandleRoleJoinTable()
    {
        if (!m_IsLoadInfo)
            return;
        m_IsInMonth = false ;
        m_LogJoinTableTime = Utility.GetTickCount();
    }
    public void SetRoleJoinMonth()
    {
        m_IsInMonth = true;
    }
    public void OnHandleRoleLeaveTable()
    {
        if (!m_IsLoadInfo)
            return;
	    //当用户离开桌子的时候 
	    //1.判断玩家是否离开比赛的桌子 并且 判断玩家在桌子上的游戏时间 用于统计数据
        UInt32 GameSec = (Utility.GetTickCount() - m_LogJoinTableTime) / 1000;
        m_GameData.TotalGameSec += GameSec;
	    if (!m_IsInMonth)
            m_GameData.NonMonthGameSec += GameSec;
    }
    public void OnHandleRoleSignUpMonth()
    {
        if (!m_IsLoadInfo)
            return;
	    //玩家报名参加比赛
        m_GameData.RoleMonthSigupSum += 1;
    }
    public void OnHandleRoleCombo(ushort ComboCount)
    {
        if (!m_IsLoadInfo)
            return;
        UInt16 Combo = 0;
        if (m_GameData.MaxCombSum >= Combo)
            return;
        m_GameData.MaxCombSum = Combo;
    }


    private void SendLoadRoleGameData()
    {
        CL_Cmd_GetGameData ncb = new CL_Cmd_GetGameData();
        ncb.SetCmdType(NetCmdType.CMD_CL_GetGameData);
        NetServices.Instance.Send<CL_Cmd_GetGameData>(ncb);
    }
    public bool HandleCmd(NetCmdBase obj)
    {
        switch (obj.GetCmdType())
        {
            case NetCmdType.CMD_LC_GendGameData:
                return HandleLoadRoleGameData(obj);
        }
        return true;
    }
    bool HandleLoadRoleGameData(NetCmdBase obj)
    {
        LC_Cmd_GendGameData ncb = (LC_Cmd_GendGameData)obj;
        m_GameData = ncb.GameData;
        m_IsLoadInfo = true;
        tagGameDataChangeEvent pEvent = new tagGameDataChangeEvent();
        MsgEventHandle.HandleMsg(pEvent);
        return true;
    }
}